com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs

```diff
@@ -237,19 +237,32 @@ public static ObservationProto GetObservationProto(this ISensor sensor, Observat
             var shape = sensor.GetObservationShape();
             ObservationProto observationProto = null;
             var compressionType = sensor.GetCompressionType();
-            // Check capabilities if we need to concatenate PNGs
-            if (compressionType == SensorCompressionType.PNG && shape.Length == 3 && shape[2] > 3)
+            if (!s_HaveWarnedAboutTrainerCapabilities)
             {
-                var trainerCanHandle = Academy.Instance.TrainerCapabilities == null || Academy.Instance.TrainerCapabilities.ConcatenatedPngObservations;
-                if (!trainerCanHandle)
+                // Check capabilities if we need to concatenate PNGs
+                if (compressionType == SensorCompressionType.PNG && shape.Length == 3 && shape[2] > 3)
                 {
-                    if (!s_HaveWarnedAboutTrainerCapabilities)
+                    var canHandleMultiPNG = Academy.Instance.TrainerCapabilities == null || Academy.Instance.TrainerCapabilities.ConcatenatedPngObservations;
+                    if (!canHandleMultiPNG)
                     {
                         Debug.LogWarning($"Attached trainer doesn't support multiple PNGs. Switching to uncompressed observations for sensor {sensor.GetName()}.");
-                        s_HaveWarnedAboutTrainerCapabilities = true;
+                        compressionType = SensorCompressionType.None;
                     }
-                    compressionType = SensorCompressionType.None;
                 }
+
+                // Check capabilities if we need mapping for compessed observations
+                if (compressionType != SensorCompressionType.None)
+                {
+                    var canHandleMapping = Academy.Instance.TrainerCapabilities == null || Academy.Instance.TrainerCapabilities.CompressionMappings;
+                    if (!canHandleMapping)
+                    {
+                        Debug.LogWarning("Attached trainer doesn't support compression mapping. " +
```

Comment:
A warning is neither sufficient nor always necessary here. You should warn if the trainer can't handle it (no "potentially") and fall back to uncompressed observations, otherwise you shouldn't warn and should keep using compression.

The question is what's a mapping that the trainer can't handle. I think the ones that it can handle are:
1) null mapping (either not an `ICompressibleSensor` or `GetCompressionMapping()` returns null)
2) grayscale: mapping is all 0's
3) identity mapping: mapping[i] == i for all i

You should check for these cases here (or in a helper method) and not warn/downgrade if they're true.